certbot-nginx/certbot_nginx/configurator.py

```diff
@@ -867,3 +875,18 @@ def install_ssl_options_conf(options_ssl):
     # Check to make sure options-ssl.conf is installed
     if not os.path.isfile(options_ssl):
         shutil.copyfile(constants.MOD_SSL_CONF_SRC, options_ssl)
+        return
+    # there's already a file there. if it exactly matches a previous file hash,
+    # we can update it. otherwise, create a .new and print a warning.
+    digest = crypto_util.sha256sum(options_ssl)
+    if digest in PREVIOUS_SSL_OPTIONS_HASHES: # safe to update
+        shutil.copyfile(constants.MOD_SSL_CONF_SRC, options_ssl)
+    elif digest == CURRENT_SSL_OPTIONS_HASH: # already up to date
+        return
+    else: # not safe to update
+        new_filename = options_ssl + ".new"
+        shutil.copyfile(constants.MOD_SSL_CONF_SRC, new_filename)
+        # print warning
+        logger.warning("%s has been manually modified; updated ssl configuration options "
```

Comment:
> Sticking a hash somewhere when we write and writing .new out immediately would have the same effect, I don't actually have a preference between them. I'm happy to write a hash out to some file somewhere.

If you write out the `.new` file at the same time as you write the live file it has the same effect. I have a preference for either including a special value/version in the file itself or creating an effectively hidden file containing the original contents/hash than to create the publicly visible `.new` file initially. If we did a hidden file, it could be an actually hidden file in `/etc/letsencrypt` or something placed somewhere like `/var/lib/letsencrypt` (although the latter opens us up to other problems I identified above).

 > Also, it occurs to me that we don't actually have to write out a .new -- we could just point them to constants.MOD_SSL_CONF_SRC.

Yeah that'd be fine too.

> I really don't think we should make lots of noise over and over until they tell us to shut up. Especially because they might update the contents of a field we changed, but have the file still be different from the updated version. Then, they'd have to tell us with a field or flag set somewhere, which I definitely don't want to be adding.

Very good point. Forcing the user to either take our file as is or turn off updates entirely is a bad idea. Basically we'd want some way to know the user is happy with the current state of the file, but I don't think there's really a clean way to do that. `--nginx-no-tls-updates-until-next-time` is pretty ridiculous.